classes/local/store/azure_blob_storage/client.php

```diff
@@ -0,0 +1,296 @@
+        }
+
+        $this->api = new api($config->azure_accountname, $config->azure_container, $config->azure_sastoken);
+        $this->config = $config;
+        $this->maxupload = api::MAX_BLOCK_SIZE;
+    }
+
+    /**
+     * Determines if this filesystem is available for use.
+     * @return bool
+     */
+    public function get_availability(): bool {
+        // Requires local_azureblobstorage to be installed.
+        // Namespace changed in 4.4+.
+        if (class_exists('\core\plugin_manager')) {
+            $info = \core\plugin_manager::instance()->get_plugin_info('local_azureblobstorage');
+            // For 4.2 and 4.3.
+        } else if (class_exists('\core_plugin_manager')) {
+            $info = \core_plugin_manager::instance()->get_plugin_info('local_azureblobstorage');
+        } else {
+            throw new coding_exception("Could not load plugin manager class");
+        }
+
+        // Info is empty if plugin is not installed or no API is setup (missing config?).
+        return !empty($info);
+    }
+
+    /**
+     * Sets the StreamWrapper to allow accessing the remote content via a blob:// path.
+     */
+    public function register_stream_wrapper() {
+        if ($this->get_availability()) {
+            stream_wrapper::register($this->api);
+        } else {
+            parent::register_stream_wrapper();
+        }
+    }
+
+    /**
+     * Returns the full path for a given file by contenthash
+     * @param string $contenthash
+     * @return string filepath
+     */
+    public function get_fullpath_from_hash($contenthash): string {
+        $filepath = $this->get_filepath_from_hash($contenthash);
+        return "blob://$filepath";
+    }
+
+    /**
+     * Returns the filepath from the contenthash, mimicking the
+     * structure of the filedir storage system.
+     * @param string $contenthash
+     * @return string filepath
+     */
+    protected function get_filepath_from_hash($contenthash): string {
+        $l1 = $contenthash[0] . $contenthash[1];
+        $l2 = $contenthash[2] . $contenthash[3];
+        return "$l1/$l2/$contenthash";
+    }
+
+    /**
+     * Returns the blob key (the key used to reference the blob) from a given filepath.
+     * @param string $filepath
+     * @return string
+     */
+    protected function get_blob_key_from_path(string $filepath): string {
+        return str_replace("blob://", '', $filepath);
+    }
+
+    /**
+     * Deletes a given file
+     * @param string $fullpath
+     */
+    public function delete_file($fullpath) {
+        // Stream wrapper supports unlinking, so just unlink.
+        unlink($fullpath);
+    }
+
+    /**
+     * Renames a given file
+     * @param string $currentpath
+     * @param string $destinationpath
+     */
+    public function rename_file($currentpath, $destinationpath) {
+        // Azure does not support renaming, instead the file is copied
+        // and the old one is deleted.
+        copy($currentpath, $destinationpath);
```

Comment:
We should check for success here before actually deleting the original object

Reply:
Done